docs/gguf.md

```diff
@@ -0,0 +1,372 @@
+# GGUF
+
+GGUF is a file format for storing models for inference with GGML and executors based on GGML. GGUF is a binary format that is designed for fast loading and saving of models, and for ease of reading. Models are traditionally developed using PyTorch or another framework, and then converted to GGUF for use in GGML.
+
+It is a successor file format to GGML, GGMF and GGJT, and is designed to be unambiguous by containing all the information needed to load a model. It is also designed to be extensible, so that new features can be added to GGML without breaking compatibility with older models.
+
+For more information about the motivation behind GGUF, see [Current State of Affairs](#current-state-of-affairs).
+
+## Specification
+
+GGUF is a format based on the existing GGJT, but makes a few changes to the format to make it more extensible and easier to use. The following features are desired:
+
+- Single-file deployment: they can be easily distributed and loaded, and do not require any external files for additional information.
+- Extensible: new features can be added to GGML without breaking compatibility with existing models.
+- `mmap` compatibility: models can be loaded using `mmap` for fast loading and saving.
+- Easy to use: models can be easily loaded and saved using a small amount of code, with no need for external libraries, regardless of the language used.
+- Full information: all information needed to load a model is contained in the model file, and no additional information needs to be provided by the user.
+
+The key difference between GGJT and GGUF is the use of a key-value structure for the hyperparameters (now referred to as metadata), rather than a list of untyped values. This allows for new metadata to be added without breaking compatibility with existing models, and to annotate the model with additional information that may be useful for inference or for identifying the model.
+
+### File Structure
+
+GGUF files are structured as follows. They assume the use of a global `ALIGNMENT` constant, which is the alignment of the model data. This is currently 64 bytes, but may change in the future. [^1] To achieve this, where relevant, the file is padded with `0x00` bytes to the next multiple of `ALIGNMENT`.
+
+[^1]: This may be moved to a per-model key-value pair in the future.
+
+```c
+enum ggml_type {
+    GGML_TYPE_F32  = 0,
+    GGML_TYPE_F16  = 1,
```

Comment:
What about BF16 ?

Reply:
Currently if you want the highest quality you have to double the tensor sizes by using F32. My guess is that BF16 is not natively supported by many platform architectures yet. I would also like to see support for BF16 in ggml. I wonder if BF16 emulation really is be slower than F32, since it is in fact a truncated version of F32. @ggerganov ?